### PR TITLE
Highlight changes to the memory regions that are being watched.

### PR DESCRIPTION
**Pull request by trittweiler**
Configurable via:

   dashboard -style memory highlight-changes False|True|"cumulative"

False        means the feature is disabled.
True         means to highlight the changes of exactly the last gdb command.
"cumulative" means to highlight all changes, where changes of the last
             gdb command are highlighted specially.

The default is True.

The highlighted changes can be cleared by unwatching, and (re)watching
the particular memory region.

Tested with gdb 8.1.0, and Python 3.6.5.

### CODE REVIEW

**Comment by trittweiler** [comment]
Hi Andrea,

Please let me know if you are happy with the changes as they are.

In particular the configuration related bits, i.e. the `dashboard -style memory highlight-changes False|True|"cumulative"`. I would have liked the configuration not having to involve actual strings, but just be  `... highlight-changes disabled|enabled|cumulative` but that's not something the infrastructure supports right now.

Let me know!

And, of course, molte grazie for creating gdb-dashboard. :)

**Comment by cyrus-and** [comment]
Hi Tobias, thanks for this PR!

I hope to take a look at the code in the next days.

In the meantime I just gave it a try, the highlighting of the memory changes is a very desirable feature as we already have it for registers.

A couple of things though, why not making it always on? Is there a reason one would want to turn it off? It's always on for registers.

I have mixed feelings about the cumulative feature, I think that it's an interesting feature per se, but I don't particularly like that in order to reset it one has to watch-unwatch the expression; maybe adding another method (e.g., `reset`) would be more appropriate.

Also, I've been thinking that registers too may benefits of the same feature...

**Comment by trittweiler** [comment]
Hi Andrea,

Thanks for the feedback.

Please note that the commit makes sure that the feature *is* enabled by default. It always highlights the changes of the last commit - just like it's done for registers.

Adding an explicit `reset` command will be easy. Just note that the reset command will also require the specification of the address. So `reset 0xdeadbeef` versus `unwatch 0xdeadbeef` + cursor-up + C-a + M-w + `watch` isn't all that much more to type. And usually I would do a `C-r watch` to find the address in the history and then just prepend an "un" to the `watch ...`.

**Comment by cyrus-and** [comment]
> Just note that the reset command will also require the specification of the address.

You're right, leave it as it is for now. Also, I just noticed that simply re-watching it again works as well, that's enough.

**Comment by cyrus-and** [comment]
Also I'm not sure I get why you're keeping track of the "number of reads", isn't keeping the initial raw bytes and comparing with the next `inferior.read_memory(...)` enough?

**Comment by trittweiler** [comment]
Hi Andrea,

Thanks for the review. I will follow up with the changes later today.

> Also I'm not sure I get why you're keeping track of the "number of reads", isn't keeping the initial raw bytes and comparing with the next `inferior.read_memory(...)` enough?

I wanted to make changes of the very last command stand out. So they're printed in bold. (`R.style_selected_1` versus `R.style_selected_2`)

I also had the idea of adding a feature were changes are color-rotated. I didn't implement it in the end because I assumed that the number of colors are limited in a terminal, and even they aren't on some terminals I would have needed to dig deeper into TerminalFormatter than I wanted.

**Comment by trittweiler** [comment]
> Yes, but I see no value in allowing to turn this feature off. Just provide a bool cumulative-changes which defaults to False, this would also eliminate the problem of having False|True|"cumulative" as allowed values.

Feature is turned off if set to `False`
Feature is enabled, and only the effects of the very last command are highlighted when set to `True`
When set to `"cumulative"`, changes of the very last command are highlighted via `R.style_selected_1`, older changes via `R.style_selected_2`

As mentioned in the other command, I had envisioned the possibility of another value `"rotate"` here, too. 

But perhaps stuffing it all into a single config item is too cramped.

I will change as you suggested.

**Comment by cyrus-and** [comment]
> Thanks for the review. I will follow up with the changes later today.

No rush, thanks!

> I wanted to make changes of the very last command stand out. So they're printed in bold. (R.style_selected_1 versus R.style_selected_2)

That makes sense. But maybe it's enough to keep the original and the latest bytes, then compare the current with both and display accordingly.

> I also had the idea of adding a feature were changes are color-rotated.

I'd keep it simple.

**Comment by trittweiler** [comment]
Hi Andrea,

I would usually squash the later changes into the previous changeset and do a push --force, so to yield a single coherent commit.

I am not sure what github will do in that case with the PR and the (now stale) review comments.

How do you want to handle this? Just add another commit and you merge both?

**Comment by cyrus-and** [comment]
As you wish, I have the option here to squash all the commits on master yet retaining the authorship.

File: .gdbinit [hunk 1]
```diff
@@ -1372,6 +1375,78 @@ class Memory(Dashboard.Module):
...
+            }
+        }
+
+class Memdump ():
+    def __init__(self, address, length):
+        self.start_address = address
+        self.end_address = address + length
+        self.length = length
+        self.bytes = None
+        self.changes = {}       # map address to Nth iteration it changed
```
[the remark this answers is not in the document]
Comment (trittweiler) [reply .gdbinit]: I wasn't sure what you were alluding to with "lowercase" because other comments also seem to start with a lowercase character.

**Comment by PokerMakyo** [comment]
Is this going to be merged? Looks like a really nice feature.

**Comment by cyrus-and** [comment]
Sorry I kind of lost track of this... anyway I switched to a simpler implementation in 0286c1662e048d119dec671fd0fbb8533491ab81.

**Pull request closed without merging**